lib/puppet/functions/fact.rb

```diff
@@ -31,7 +31,7 @@ def to_dot_syntax(array_path)
   end
 
   def fact(fact_name)
-    facts = closure_scope['facts']
+    facts = closure_scope['facts'] || Facter.to_hash
```

Comment:
If facts aren't in scope (which are supposed to be the facts submitted by the agent), then aren't we're going to use the puppetserver's facts (?) which is bad.

Reply:
That's correct. Are there cases in which facts would not be in scope during a standard catalog compilation?

This is intended to use the in-scope facts during standard catalog compilation and fall back to running facter when deferred to the agent. Is there a better way?